python/cuda_parallel/tests/test_unique_by_key.py

```diff
@@ -0,0 +1,363 @@
+# Copyright (c) 2024-2025, NVIDIA CORPORATION & AFFILIATES. ALL RIGHTS RESERVED.
+#
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from typing import List
+
+import cupy as cp
+import numba.cuda
+import numpy as np
+import pytest
+
+import cuda.parallel.experimental.algorithms as algorithms
+import cuda.parallel.experimental.iterators as iterators
+from cuda.parallel.experimental.struct import gpu_struct
+
+DTYPE_LIST = [
+    np.uint8,
+    np.uint16,
+    np.uint32,
+    np.uint64,
+    np.int8,
+    np.int16,
+    np.int32,
+    np.int64,
+    np.float32,
+    np.float64,
+]
+
+
+def random_array(size, dtype, max_value=None) -> np.typing.NDArray:
+    rng = np.random.default_rng()
+    if np.isdtype(dtype, "integral"):
+        if max_value is None:
+            max_value = np.iinfo(dtype).max
+        return rng.integers(max_value, size=size, dtype=dtype)
+    elif np.isdtype(dtype, "real floating"):
+        return rng.random(size=size, dtype=dtype)
+    else:
+        raise ValueError(f"Unsupported dtype {dtype}")
+
+
+def type_to_problem_sizes(dtype) -> List[int]:
+    if dtype in [np.uint8, np.int8]:
+        return [2, 4, 5, 6]
+    elif dtype in [np.uint16, np.int16]:
+        return [4, 8, 14]
+    elif dtype in [np.uint32, np.int32, np.float32]:
+        return [4, 10, 20]
+    elif dtype in [np.uint64, np.int64, np.float64]:
+        return [4, 10, 20]
+    else:
+        raise ValueError("Unsupported dtype")
+
+
+def unique_by_key_device(
+    d_in_keys,
+    d_in_items,
+    d_out_keys,
+    d_out_items,
+    d_out_num_selected,
+    op,
+    num_items,
+    stream=None,
+):
+    unique_by_key = algorithms.unique_by_key(
+        d_in_keys, d_in_items, d_out_keys, d_out_items, d_out_num_selected, op
+    )
+
+    temp_storage_size = unique_by_key(
+        None,
+        d_in_keys,
+        d_in_items,
+        d_out_keys,
+        d_out_items,
+        d_out_num_selected,
+        num_items,
+        stream=stream,
+    )
+    d_temp_storage = numba.cuda.device_array(
+        temp_storage_size, dtype=np.uint8, stream=stream.ptr if stream else 0
+    )
+    unique_by_key(
+        d_temp_storage,
+        d_in_keys,
+        d_in_items,
+        d_out_keys,
+        d_out_items,
+        d_out_num_selected,
+        num_items,
+        stream=stream,
+    )
+
+
+def is_equal_func(lhs, rhs):
+    return lhs == rhs
+
+
+def unique_by_key_host(keys, items, is_equal=is_equal_func):
+    # Must implement our own version of unique_by_key since np.unique() returns
+    # unique elements across the entire array, while cub::UniqueByKey
+    # de-duplicates consecutive keys that are equal.
```

Comment:
Just a note that this functionality is implemented by [itertools.groupby](https://docs.python.org/3/library/itertools.html#itertools.groupby), where doing a unique-by-key would be something like:

```python
def unique_by_key(keys, vals):
    data = zip(keys, vals)
    uniqs = []
    for k, g in itertools.groupby(data, key=lambda x: x[0]):
        _, group_vals = zip(*g)
        uniqs.append(list(set(group_vals)))
    return uniqs
```

Reply:
I would prefer to keep the current implementation because I feel like it is more readable, but wouldn't mind changing it either